### PR TITLE
CAMEL-20166 : Create component Elasticsearch Rest Client

### PR DESCRIPTION
**Pull request by zbendhiba**
This is a Preview version, and I plan  support for next LTS

### CODE REVIEW

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+                            // Get response
+                            String responseBody = EntityUtils.toString(response.getEntity());
+                            // Create a Json Object from the response
+                            JsonObject jsonObject = convertHttpEntityToJsonObject(responseBody);
+                            populateExchange(jsonObject);
+                        } catch (IOException e) {
+                            exchange.setException(e);
+                        }
+
+                        callback.done(true);
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:155]: should be done(false)

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+
+    public ElasticsearchRestClientProducer(ElasticsearchRestClientEndpoint endpoint) {
+        super(endpoint);
+        this.endpoint = endpoint;
+    }
+
+    @Override
+    public boolean process(Exchange exchange, AsyncCallback callback) {
+        // getting configuration from Endpoint
+        ElasticsearchRestClientOperation operation = this.endpoint.getOperation();
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:76]: Better to validate this in component when creating endpoint, or in endpoint when creating proudcer, or in this class in doInit method

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        if (operation == null) {
+            LOG.error("Operation value is mandatory");
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:78]: Do not log and throw exception, just throw excpetion.

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+            throw new IllegalArgumentException(
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:79]: however its wrong to throw exception from the async method you use in Camel.
Instead you need to set the exception on exchange, and call calldone.done(true) and return true.

So somtimes its better to make a 

doProcess that can throw eception and call this from processing in a try .. .. catch

process
   try 
      return doProcess
   catch
      exchange set exception
       callback done true
       return true

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+            indexName = exchange.getMessage().getHeader(ElasticSearchRestClientConstant.INDEX_NAME, String.class);
+            if (ObjectHelper.isEmpty(indexName)) {
+                LOG.error("Index Name is mandatory");
+                throw new IllegalArgumentException(
+                        "Index Name is mandatory");
+            }
+        }
+
+        // get or create the Rest Client
+        RestClient restClient = getorCreateRestClient();
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:93]: better to create client in doStart

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+                new ResponseListener() {
+                    @Override
+                    public void onSuccess(Response response) {
+                        try {
+                            // Get response
+                            String responseBody = EntityUtils.toString(response.getEntity());
+                            // Create a Json Object from the response
+                            JsonObject jsonObject = convertHttpEntityToJsonObject(responseBody);
+                            populateExchange(jsonObject);
+                        } catch (IOException e) {
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:151]: Better to catch Exception in general

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+                            case DELETE -> exchange.getMessage().setBody(extractDeleted(doc));
+                            case GET_BY_ID -> exchange.getMessage().setBody(extractDocument(doc));
+                            case SEARCH -> exchange.getMessage().setBody(extractSearch(doc));
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        exchange.setException(e);
+                        callback.done(true);
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:187]: done false

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+             syntax = "elasticsearch-rest-client:clusterName", producerOnly = true,
+             category = { Category.SEARCH })
+public class ElasticsearchRestClientEndpoint extends DefaultEndpoint {
+    @UriPath
+    @Metadata(required = true)
+    private String clusterName;
+    @UriParam
+    ElasticsearchRestClientOperation operation;
+
+    @UriParam(label = "advanced")
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java:49]: autowired = true, label = advanced

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    RestClient restClient;
+
+    @UriParam
+    String indexName;
+
+    @UriParam
+    List<HttpHost> hostAddressesList;
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java:56]: Its more tooling friendly to have this as a string with multiple hosts separated by comma

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+    @UriParam(defaultValue = "30000")
+    private int connectionTimeout;
+
+    @UriParam(defaultValue = "30000")
+    private int socketTimeout;
+
+    @UriParam(label = "security")
+    private String user;
+    @UriParam(label = "security")
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java:66]: secret = true for username and password

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+    private String password;
+    @UriParam(label = "security")
+    @Metadata(supportFileReference = true)
+    private String certificatePath;
+
+    @UriParam(label = "advanced")
+    private boolean enableSniffer;
+
+    @UriParam(label = "advanced", defaultValue = "60000")
+    private int snifferInterval;
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java:76]: the int is 0 by default, so maybe it should also be 60000

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    @UriParam(label = "advanced", defaultValue = "60000")
+    private int sniffAfterFailureDelay;
```
Comment (davsclaus) [inline components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientEndpoint.java:78]: the int is 0 by default, so maybe it should also be 60000

File: components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java (hunk 4 above)
Comment (zbendhiba) [reply components/camel-elasticsearch-rest-client/src/main/java/org/apache/camel/component/elasticsearch/rest/client/ElasticsearchRestClientProducer.java:79]: @davsclaus Do you have any example for this one?

**Comment by zbendhiba** [comment]
thanks @davsclaus for the review. I'm working on the suggested updates and I'll push the changes shortly.

**Comment by zbendhiba** [comment]
@davsclaus I pushed a new version, and I think I got all the suggested updates.

**Review by oscerd — APPROVED** [review]

**Review by davsclaus — APPROVED** [review]

**Comment by davsclaus** [comment]
The doc could use a bit more examples and polish, but you can do that later in other PRs. So lets merge this so we can get a new 4.3 release out

**Pull request merged**